### PR TITLE
Fixed issue #5 as well as two other inconsistencies

### PR DESCRIPTION
**Pull request by gentledepp**
1) Fixed issue #5 (Mvx.IocConstruct<type>() only resolves the type but should register it implicitly)
2) Also, Autofacs DependencyResolutionExceptions are caught and wrapped into an MvxIoCResolveException
3) Finally, propertyinjection can be configured to inject into all properties or just those marked with MvxInject as forseen in MvvmCross

### CODE REVIEW

File: nuget/Autofac.Extras.MvvmCross.nuspec [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (alexmg) [inline nuget/Autofac.Extras.MvvmCross.nuspec]: This information should be taken from the project.json file as part of the build. Why is this .nuspec file required?

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs [hunk 2]
```diff
@@ -634,19 +683,43 @@ public virtual bool TryResolve(Type type, out object resolved)
                 throw new ArgumentNullException(nameof(registration));
             }
 
-            if (this.PropertyInjectorOptions.InjectIntoProperties == MvxPropertyInjection.None)
+            var options = this.PropertyInjectionOptions as IAutofacPropertyInjectorOptions;
```
Comment (alexmg) [inline src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:686]: Looks like the `IPropertySelector` may have been better passed as a separate parameter in the previous implementation. I'm not sure how the `IMvxPropertyInjectorOptions` are supposed to fit into the DI abstraction as the `IMvxIoCProvider` doesn't reference them directly.

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs [hunk 3]
```diff
@@ -42,18 +44,36 @@ namespace Autofac.Extras.MvvmCross
...
+            if (propertyInjectionOptions == null)
+                throw new ArgumentNullException(nameof(propertyInjectionOptions));
+
+            Container = container;
+            PropertyInjectionOptions = propertyInjectionOptions;
+            PropertyInjectionEnabled = propertyInjectionOptions.InjectIntoProperties != MvxPropertyInjection.None;
+
+            if (propertyInjectionOptions.ThrowIfPropertyInjectionFails)
+            {
+                throw new NotSupportedException("AutoFac does not support throwing an exception in case a service could not be injected into a property!");
```
Comment (alexmg) [inline src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs]: Nit. Correct casing is Autofac. 😃

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs [hunk 4]
```diff
@@ -387,7 +403,8 @@ public virtual void RegisterSingleton(Type tInterface, object theObject)
 
             // You can't inject properties on a pre-constructed instance.
             cb.RegisterInstance(theObject).As(tInterface).AsSelf().SingleInstance();
-            cb.Update(this.Container);
+
+            cb.Update(Container);
```
Comment (alexmg) [inline src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:407]: Is there is a way to configure all the registrations up front using a container builder and then build the container in a single step? I don't see anything indicating such support in the `IMvxIoCProvider` interface.

Comment (alexmg) [reply src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:407]: I know you didn't add this code, just hoping you might have the answer.

File: test/Autofac.Extras.MvvmCross.Test/AutofacMvxTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace Autofac.Extras.MvvmCross.Test
+{
+    /// <summary>
+    /// Base class for all tests, as the AutofacMvxIocProvider must be an "MvxSingleton"
+    /// Multiple tests running in parallel would cause multiple providers to be instanciated leading to the following exception:
+    /// MvvmCross.Platform.Exceptions.MvxException : You cannot create more than one instance of MvxSingleton
+    /// </summary>
+    public abstract class AutofacMvxTestBase : IDisposable
+    {
+        private static readonly AsyncLock _lock = new AsyncLock();
```
Comment (alexmg) [inline test/Autofac.Extras.MvvmCross.Test/AutofacMvxTestBase.cs:13]: A shame this is required as it negates the benefit of a parallel run.

**Comment by alexmg** [comment]
Glad you took the time to prep another PR. Thanks for taking the effort.

File: nuget/Autofac.Extras.MvvmCross.nuspec (hunk 1 above)
Comment (gentledepp) [reply nuget/Autofac.Extras.MvvmCross.nuspec]: Ok - so I missed that. Sry. 
Did not know that you can nowadays generate the nuget package directly from the information within the project.json.

How does this work? (can you share a link please?)

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs (hunk 4 above)
Comment (gentledepp) [reply src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:407]: The reason is that you register instances in MvvmCross using "Mvx.IocRegisterSingleton".
So you never really see the underlying DI frameworks internals which is quite good I think

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs (hunk 2 above)
Comment (gentledepp) [reply src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:686]: They fit as you wire-up your custom DI framework in mvvmcross once when the app starts up. This is the only line of code where the actual DI framework (in that case Autofac) shines through the facade of MvvmCross.

File: test/Autofac.Extras.MvvmCross.Test/AutofacMvxTestBase.cs (hunk 5 above)
Comment (gentledepp) [reply test/Autofac.Extras.MvvmCross.Test/AutofacMvxTestBase.cs:13]: Yeah, but mvvmcross requires your IMvxIocProvider to inherit MvxSingleton<IMvxIocProvider> in order to reference it from the "Mvx.Ioc.... " static methods.

**Comment by gentledepp** [comment]
yeah - sorry for beeing so grumpy yesterday. We are migrating to the latest MvvmCross version right now and this library made the most problems although it's by far the smallest. :-|

File: nuget/Autofac.Extras.MvvmCross.nuspec (hunk 1 above)
Comment (alexmg) [reply nuget/Autofac.Extras.MvvmCross.nuspec]: The core Autofac library does this. https://github.com/autofac/Autofac/blob/develop/src/Autofac/project.json

File: src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs (hunk 4 above)
Comment (alexmg) [reply src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:407]: I assumed that would be the case. I couldn't see anything of that sort with a quick look.

**Comment by alexmg** [comment]
Looks like if you can sort the project.json issue out and fix up the casing of "Autofac" this one is good to go.

**Comment by gentledepp** [comment]
I'm glad to hear that :)

**Comment by alexmg** [comment]
Sorry it took so long to get this one through the pipeline. This has now been pushed to NuGet as v4.0.2. The label wasn't consistent between the package and AppVeyor build on v4.0.1 so I jumped it up to v4.0.2 to keep them aligned.

**Pull request merged**